File: sections/14_core_patterns_mixed_contest/problems/n_wildcard_choice_map/solution.cpp
#include <bits/stdc++.h>
using namespace std;

bool intersects(int low1, int high1, int low2, int high2) {
    return max(low1, low2) <= min(high1, high2);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s;
    cin >> s;
    int n = s.size();
    vector<int> prefix_low(n + 1), prefix_high(n + 1);
    for (int i = 0; i < n; ++i) {
        if (prefix_low[i] > prefix_high[i]) {
            prefix_low[i + 1] = 1;
            prefix_high[i + 1] = 0;
            continue;
        }
        int low = prefix_low[i];
        int high = prefix_high[i];
        if (s[i] == '(') ++low, ++high;
        else if (s[i] == ')') --low, --high;
        else --low, ++high;
        int parity = (i + 1) & 1;
        if (low < 0) low = parity;
        prefix_low[i + 1] = low;
        prefix_high[i + 1] = high;
    }
    if (n % 2 || prefix_low[n] > 0 || prefix_high[n] < 0) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    vector<int> suffix_low(n + 1), suffix_high(n + 1);
    for (int i = n - 1; i >= 0; --i) {
        if (suffix_low[i + 1] > suffix_high[i + 1]) {
            suffix_low[i] = 1;
            suffix_high[i] = 0;
            continue;
        }
        int low = suffix_low[i + 1];
        int high = suffix_high[i + 1];
        if (s[i] == '(') {
            low = low == 0 ? 1 : low - 1;
            high -= 1;
        } else if (s[i] == ')') {
            ++low;
            ++high;
        } else {
            low = low == 0 ? 1 : low - 1;
            ++high;
        }
        suffix_low[i] = low;
        suffix_high[i] = high;
    }

    string answer = s;
    for (int i = 0; i < n; ++i) {
        if (s[i] != '?') continue;
        bool can_open = intersects(
            prefix_low[i] + 1, prefix_high[i] + 1,
            suffix_low[i + 1], suffix_high[i + 1]);
        int closing_low = max(prefix_low[i], 1) - 1;
        bool can_close = intersects(
            closing_low, prefix_high[i] - 1,
            suffix_low[i + 1], suffix_high[i + 1]);
        if (can_open && !can_close) answer[i] = '(';
        else if (!can_open && can_close) answer[i] = ')';
        else answer[i] = '?';
    }
    cout << answer << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/n_wildcard_choice_map/solution.py
import sys


def main() -> None:
    s = sys.stdin.buffer.readline().strip().decode()
    n = len(s)
    prefix_low = [0] * (n + 1)
    prefix_high = [0] * (n + 1)
    for i, char in enumerate(s):
        if prefix_low[i] > prefix_high[i]:
            prefix_low[i + 1], prefix_high[i + 1] = 1, 0
            continue
        low, high = prefix_low[i], prefix_high[i]
        if char == "(":
            low += 1
            high += 1
        elif char == ")":
            low -= 1
            high -= 1
        else:
            low -= 1
            high += 1
        if low < 0:
            low = (i + 1) & 1
        prefix_low[i + 1] = low
        prefix_high[i + 1] = high

    if n % 2 or not prefix_low[n] <= 0 <= prefix_high[n]:
        print("IMPOSSIBLE")
        return

    suffix_low = [0] * (n + 1)
    suffix_high = [0] * (n + 1)
    for i in range(n - 1, -1, -1):
        if suffix_low[i + 1] > suffix_high[i + 1]:
            suffix_low[i], suffix_high[i] = 1, 0
            continue
        low, high = suffix_low[i + 1], suffix_high[i + 1]
        if s[i] == "(":
            low = 1 if low == 0 else low - 1
            high -= 1
        elif s[i] == ")":
            low += 1
            high += 1
        else:
            low = 1 if low == 0 else low - 1
            high += 1
        suffix_low[i], suffix_high[i] = low, high

    answer = list(s)
    for i, char in enumerate(s):
        if char != "?":
            continue
        after_low, after_high = suffix_low[i + 1], suffix_high[i + 1]
        can_open = max(prefix_low[i] + 1, after_low) <= min(
            prefix_high[i] + 1, after_high)
        closing_low = max(prefix_low[i], 1) - 1
        can_close = max(closing_low, after_low) <= min(
            prefix_high[i] - 1, after_high)
        if can_open and not can_close:
            answer[i] = "("
        elif can_close and not can_open:
            answer[i] = ")"
    print("".join(answer))


if __name__ == "__main__":
    main()
